watchtower/wtclient/session_queue.go

```diff
@@ -0,0 +1,512 @@
+			conn, stateUpdate, q.localInit, sendInit,
+		)
+		if err != nil {
+			log.Errorf("Unable to send state update: %v", err)
+
+			// Always apply a small delay between sends,
+			// which makes the unit tests more reliable. If we were
+			// requested to back off, when we will do so.
+			delay := time.Millisecond
+			if shouldBackoff {
+				// TODO(conner): add reb
+				delay = retryInterval
+			}
+
+			select {
+			case <-time.After(delay):
+			case <-q.forceQuit:
+			}
+			return
+		}
+
+		// If the last task was backed up successfully, we'll exit and
+		// continue once more tasks are added to the queue.
+		if isLast {
+			return
+		}
+
+		// If the first state update was sent successfully, we can skip
+		// sending an Init messages for subsequent updates.
+		sendInit = false
+	}
+}
+
+// sendStateUpdate sends a wtwire.StateUpdate to the watchtower and processes
+// the ACK before returning. If sendInit is true, this method will first send
+// the localInit message and verify that the tower supports our required feature
+// bits. And error is returned if any part of the send fails. The boolean return
+// variable indicates whether or not we should back off before attempting to
+// send the next state update.
+func (q *sessionQueue) sendStateUpdate(conn wtserver.Peer,
+	stateUpdate *wtwire.StateUpdate, localInit *wtwire.Init,
+	sendInit bool) (bool, error) {
+
+	// If this is the first message being sent to the tower, we must send an
+	// Init message to establish that server supports the features we
+	// require.
+	if sendInit {
+		// Send Init to tower.
+		err := q.cfg.SendMessage(conn, q.localInit)
+		if err != nil {
+			return false, err
+		}
+
+		// Receive Init from tower.
+		remoteMsg, err := q.cfg.ReadMessage(conn)
+		if err != nil {
+			return false, err
+		}
+
+		remoteInit, ok := remoteMsg.(*wtwire.Init)
+		if !ok {
+			return true, err
+		}
+
+		// Validate Init.
+		err = q.localInit.CheckRemoteInit(
+			remoteInit, wtwire.FeatureNames,
+		)
+		if err != nil {
+			return true, err
+		}
+	}
+
+	// Send StateUpdate to tower.
+	err := q.cfg.SendMessage(conn, stateUpdate)
+	if err != nil {
+		return false, err
+	}
+
+	// Receive StateUpdate from tower.
+	remoteMsg, err := q.cfg.ReadMessage(conn)
+	if err != nil {
+		return false, err
+	}
+
+	stateUpdateReply, ok := remoteMsg.(*wtwire.StateUpdateReply)
+	if !ok {
+		return true, err
+	}
+
+	// Process the reply from the tower.
+	switch stateUpdateReply.Code {
+
+	// The tower reported a successful update, validate the response and
+	// record the last applied returned.
+	case wtwire.CodeOK:
+		lastApplied := stateUpdateReply.LastApplied
+		if lastApplied > stateUpdate.SeqNum {
+			// TODO(conner): borked watchtower?
+		}
+
+		if lastApplied < q.cfg.ClientSession.TowerLastApplied {
+			// TODO(conner): borked watchtower?
+		}
+
+		// TODO(conner): store last applied for session
+		// TODO(conner): delete blob locally
+		q.queueCond.L.Lock()
+		q.cfg.ClientSession.SeqNum++
```

Comment:
When does this state get updated on disk?

Reply:
latest version properly updates the db, you're correct that the first version did this assignment in memory ;)